pageserver/src/layered_repository.rs

```diff
@@ -1638,34 +1508,273 @@ impl LayeredTimeline {
+            for (img_range, last_img) in image_coverage {
+                let img_lsn = if let Some(ref last_img) = last_img {
+                    last_img.get_lsn_range().end
+                } else {
+                    Lsn(0)
+                };
+
+                let num_deltas = layers.count_deltas(&img_range, &(img_lsn..lsn))?;
+
+                info!(
+                    "range {}-{}, has {} deltas on this timeline",
+                    img_range.start, img_range.end, num_deltas
+                );
+                if num_deltas >= threshold {
+                    return Ok(true);
+                }
+            }
+        }
+
+        Ok(false)
+    }
+
+    fn create_image_layer(&self, partition: &KeySpace, lsn: Lsn) -> Result<()> {
+        let img_range =
+            partition.ranges.first().unwrap().start..partition.ranges.last().unwrap().end;
+        let mut image_layer_writer =
+            ImageLayerWriter::new(self.conf, self.timelineid, self.tenantid, &img_range, lsn)?;
+
+        for range in &partition.ranges {
+            let mut key = range.start;
+            while key < range.end {
+                let img = self.get(key, lsn)?;
+                image_layer_writer.put_image(key, &img)?;
+                key = key.next();
+            }
+        }
+        let image_layer = image_layer_writer.finish()?;
+
+        // Sync the new layer to disk before adding it to the layer map, to make sure
+        // we don't garbage collect something based on the new layer, before it has
+        // reached the disk.
+        //
+        // We must also fsync the timeline dir to ensure the directory entries for
+        // new layer files are durable
+        //
+        // Compaction creates multiple image layers. It would be better to create them all
+        // and fsync them all in parallel.
+        par_fsync::par_fsync(&[
+            image_layer.path(),
+            self.conf.timeline_path(&self.timelineid, &self.tenantid),
+        ])?;
+
+        // FIXME: Do we need to do something to upload it to remote storage here?
 
-        let mut layer_paths = Vec::new();
-        let _write_guard = self.write_lock.lock().unwrap();
         let mut layers = self.layers.lock().unwrap();
+        layers.insert_historic(Arc::new(image_layer));
+        drop(layers);
 
-        // Finally, replace the frozen in-memory layer with the new on-disk layers
-        layers.remove_historic(layer);
+        Ok(())
+    }
 
-        // Add the historics to the LayerMap
-        for delta_layer in new_historics.delta_layers {
-            layer_paths.push(delta_layer.path());
-            layers.insert_historic(Arc::new(delta_layer));
+    fn compact_level0(&self, target_file_size: u64) -> Result<()> {
+        let layers = self.layers.lock().unwrap();
+
+        // We compact or "shuffle" the level-0 delta layers when 10 have
+        // accumulated.
+        static COMPACT_THRESHOLD: usize = 10;
```

Comment:
Should it be changed to parameters?
So now inmemory layer is evicted when it reaches `checkpoint_distance`  = 256Mb
And compaction is performed only once 10 layers are evicted, so we need to write ~2Gb of WAL before compaction is triggered and image layers can be generated. It seems to be too long...
Also I do not understand reasons by performing repartitioning and materialization (image layer generation) in the same thread.